reporting_app/static/charts.js

```diff
@@ -1,328 +1,547 @@
-    var i;
-    var data_len = run_elements.length;  // cache the data length to avoid looking it up 30,000 times
-    var j;
-    var fields_len = fields.length;
-    for (i=0; i<data_len; i++) {
-        var e = run_elements[i];
-        // startOf and endOf have side-effects, so these need to be done on new moment() objects
-        var st = moment(e['date']).startOf(time_period);
-        var en = moment(e['date']).endOf(time_period);
-        var middle_of_time_period = st.add(en.diff(st) / 2);
-
-        if (aggregate[middle_of_time_period] == undefined) {
-            aggregate[middle_of_time_period] = {};
-            for (j=0; j<fields_len; j++) {
-                aggregate[middle_of_time_period][fields[j]] = e[fields[j]];
-            }
-        } else {
-            for (j=0; j<fields_len; j++) {
-                aggregate[middle_of_time_period][fields[j]] += e[fields[j]];
-            }
-        }
+var colour_palette = Highcharts.getOptions().colors;
+
+//get any percentile from an array
+function getPercentile(data, percentile) {
+    data.sort(numSort);
+    var index = (percentile / 100) * data.length;
+    var result;
+    if (Math.floor(index) == index) {
+        result = (data[(index - 1)] + data[index]) / 2;
+    } else {
+        result = data[Math.floor(index)];
     }
-    return aggregate
+    return result;
 }
 
-function unwind_samples_sequenced(run_elements) {
-    var all_sample_data = [];
-    var seen_sample_ids = {};
-
-    var i;
-    var len = run_elements.length;
-    for (i=0; i<len; i++) {
-        var e = run_elements[i];
-        var sample_id = e['sample_id'];
-        var run_id = e['run_id'];
-        if (seen_sample_ids[sample_id] == undefined) {
-            seen_sample_ids[sample_id] = [run_id];
-        } else {
-            seen_sample_ids[sample_id].push(run_id);
-        }
-    }
+//because .sort() doesn't sort numbers correctly
+function numSort(a, b) {
+  return a - b;
+}
 
-    var sample;
-    var run;
-    for (sample in seen_sample_ids) {
-        var runs = seen_sample_ids[sample];
-        var uniq_runs = _.uniq(runs);
-        uniq_runs.sort();
-        var repeat_uniq_runs = uniq_runs.slice(1);
-
-        if (uniq_runs.length > 0) {
-            all_sample_data.push(
-                {
-                    'date': moment(uniq_runs[0].split('_')[0], 'YYMMDD').toDate(),
-                    'total': 1,
-                    'first': 1,
-                    'repeat': 0
-                }
-            );
-            for (run in repeat_uniq_runs) {
-                all_sample_data.push(
-                    {
-                        'date': moment(repeat_uniq_runs[run].split('_')[0], 'YYMMDD').toDate(),
-                        'total': 1,
-                        'first': 0,
-                        'repeat': 1
+function aggregate(list_object, toGroup, toAggregate, fn, output_field, val0) {
```

Comment:
It's not obvious how `toAggregate`, `fn`, `output_field` and `val0` are handled. Like in the call below:
```javascript
aggregate(a_list, 'y', ['x', 'x'], [average, count], ['average_x', 'count_x'], null);
```
the first `x` corresponds to the `average` and `average_x`, and the second `x` corresponds to the `count` and `count_x` - would be good to document this

Reply:
I've added some comment clarifying the fact that if they are arrays they need to be the same length